kotlin/internal/jvm/compile.bzl

```diff
@@ -237,81 +267,174 @@ def kt_jvm_compile_action(ctx, rule_kind, output_jar, compile_jar):
-                java_toolchain = toolchains.java,
-                host_javabase = toolchains.java_runtime,
-            )
-            compile_jar = ctx.actions.declare_file(ctx.label.name + ".abi.jar")
-            _fold_jars_action(
-                ctx,
-                rule_kind = rule_kind,
-                output_jar = compile_jar,
-                action_type = "Abi",
-                input_jars = [
-                    kt_compile_jar,
-                    java_common.run_ijar(
-                        ctx.actions,
-                        target_label = ctx.label,
-                        jar = java_compile_jar,
-                        java_toolchain = toolchains.java,
-                    ),
-                ],
-            )
+            kt_compile_jar = kt_runtime_jar
+            outputs = {
+                "output": kt_runtime_jar,
+            }
 
-    _run_kt_builder_action(
-        ctx = ctx,
+        _run_kt_builder_action(
+            ctx = ctx,
+            rule_kind = rule_kind,
+            toolchains = toolchains,
+            dirs = dirs,
+            srcs = srcs,
+            generated_src_jars = generated_src_jars,
+            friend = friend,
+            compile_deps = compile_deps,
+            annotation_processors = [],
+            transitive_runtime_jars = transitive_runtime_jars,
+            plugins = plugins,
+            outputs = outputs,
+            build_java = False,
+            build_kotlin = True,
+            mnemonic = "KotlinCompile",
+        )
+
+        compile_jars.append(kt_compile_jar)
+        output_jars.append(kt_runtime_jar)
+        if not annotation_processors:
+            kt_stubs_for_java.append(JavaInfo(compile_jar=kt_compile_jar, output_jar=kt_runtime_jar, neverlink = True))
+
+        kt_java_info = JavaInfo(
+           output_jar = kt_runtime_jar,
+           compile_jar = kt_compile_jar,
+           deps = compile_deps.deps,
+           runtime_deps = [d[JavaInfo] for d in ctx.attr.runtime_deps],
+           exports = [d[JavaInfo] for d in getattr(ctx.attr, "exports", [])],
+           neverlink = getattr(ctx.attr, "neverlink", False),
+        )
+        java_infos.append(kt_java_info)
+
+    # Build Java
+    # If there is Java source or KAPT generated Java source compile that Java and fold it into
+    # the final ABI jar. Otherwise just use the KT ABI jar as final ABI jar.
+    if srcs.java or generated_src_jars:
+        toolchain = ctx.toolchains[_TOOLCHAIN_TYPE]
+        javac_opts = [
+            # Disable option linting, it will complain about the source.
+            "-Xlint:-options",
+            "-source " + toolchain.jvm_target,
+            "-target " + toolchain.jvm_target
+        ]
+
+        # Kotlin takes care of annotation processing.
+        if len(srcs.kt) > 0:
+            javac_opts += ["-proc:none"]
+
+        java_info = java_common.compile(
+            ctx,
+            source_files = srcs.java,
+            source_jars = generated_src_jars,
```

Comment:
As mentioned earlier, maybe we can pass the compiled KAPT stub classes on the classpath rather than the KAPT generated source.

Reply:
Last I checked kapt compiled stubs aren't particularly valid bytecode. They provide just enough to run the annotation processors. 

Double check by dumping the stub bytecode, of course.

(I would keep them out of the classpath, on the principle that a leak is ugly.)

I *believe* those generated sources are the only output from the annotation processing. I could be mis-remembering how it works.